language/move-binary-format/src/serializer.rs

```diff
@@ -720,13 +723,34 @@ fn serialize_ability_sets(binary: &mut BinaryData, sets: &[AbilitySet]) -> Resul
 /// - `CodeUnit.max_stack_size` as a ULEB128
 /// - `CodeUnit.locals` as a ULEB128 (index into the `LocalSignaturePool`)
 /// - `CodeUnit.code` as variable size byte stream for the bytecode
-fn serialize_code_unit(binary: &mut BinaryData, code: &CodeUnit) -> Result<()> {
+fn serialize_code_unit(major_version: u32, binary: &mut BinaryData, code: &CodeUnit) -> Result<()> {
     serialize_signature_index(binary, &code.locals)?;
-    serialize_code(binary, &code.code)
+    serialize_code(major_version, binary, &code.code)
 }
 
 /// Serializes a single `Bytecode` instruction.
-fn serialize_instruction_inner(binary: &mut BinaryData, opcode: &Bytecode) -> Result<()> {
+fn serialize_instruction_inner(
+    major_version: u32,
+    binary: &mut BinaryData,
+    opcode: &Bytecode,
+) -> Result<()> {
+    match opcode {
+        Bytecode::LdU16(_)
+        | Bytecode::LdU32(_)
+        | Bytecode::LdU256(_)
+        | Bytecode::CastU16
+        | Bytecode::CastU32
+        | Bytecode::CastU256
+            if (major_version < VERSION_6) =>
+        {
+            return Err(anyhow!(
+                "Loading or casting u16, u32, u256 integers not supported in bytecode version {}",
+                major_version
+            ));
+        }
```

Comment:
Hmm... I'm wondering if the serializer needs to be versioned or not.

@tnowacki could you remind me what the contract here is? Do only support serializing to the latest version?

Reply:
@wrwg added this relatively recently. We currently can target serialization formats. But only V5, and now V6, are supported.

We can/will use this in the compiler for a rough and dirty ability to target old file format versions with the most recent compiler.

Reply:
Sounds good then.